optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintMatch.java

```diff
@@ -28,6 +27,16 @@
 /**
  * Retrievable from {@link ConstraintMatchTotal#getConstraintMatchSet()}
  * and {@link Indictment#getConstraintMatchSet()}.
+ *
+ * <p>
+ * Note: this class has a {@link #compareTo(ConstraintMatch)} method which is inconsistent with equals.
+ * (See {@link Comparable}.)
+ * Two different {@link ConstraintMatch} instances ({@code (a == b) == false}) must never {@link #equals(Object)},
+ * or else it would be impossible for two matches with the same justifications to coexist.
```

Comment:
I'd pull the "with the same justificaiton" list to the front, for clarity, to pin point on the case where this matters.
 I'd remove the "(a == b) == false" because "different" is clear enough.
 I'd avoid "must never ..." in favor of "can not ..." because "must" is requirements talk, while javadocs should be specification talk and clear statements on how it works.

Two different {@link ConstraintMatch} instances with the same justification list, aren't {@link equals(Object) equal}
because some ConstraintStream API methods can result in duplicate facts, which are treated as independent matches.